MIP/mip-draft-FFS/README.md

```diff
@@ -0,0 +1,229 @@
+- vote :x: for a valid block
+
+### Main challenges
+
+To achieve crypto-economically secured Fast-Finality, we need to solve the following problems:
+
+1. design a _staking_ mechanism for the validators to stake assets, distribute rewards and manage slashing
+1. _define and verify_ the threshold (e.g. 2/3 of validators attest :white_check_mark:) for L2-finality
+1. _communicate_ the L2-finality status.
+
+In addition the following may require separate discussion, as it is a different procedure (postconfirmations are handled in smart contracts on L1, whereas L2-finality is handled off-chain)
+
+4. _define and verify_ the threshold (e.g. 2/3 of validators attest :white_check_mark:) for postconfirmation.
+
+
+### Components
+
+#### Staking (addresses 1. and 2.)
+
+The staking mechanism is implemented in a contract on L1.
+This contract provides the following functionalities:
+
+- join: a new validator can join the set of validators by staking some assets
+- exit: a validator exits and  get their stakes back
+- vote: receive a vote or a set of votes, verify the integrity of the votes (signatures) and the minimum threshold (e.g. 2/3)
+
+#### Generate an L2-finality certificate (addreses 3.)
+
+To ensure that the L2-finality status is made available to third-parties, we may publish our _proof_  (2/3 of attestations :white_check_mark:) to a data availability layer and get a _certificate_ that the proof is available.  
+This DA layer should offer a reliable _mempool_ for example as described [in this paper](https://arxiv.org/pdf/2105.11827).
+
+#### Handle postconfirmations (addresses 4.)
+
+The L1 contract will verify the L2-finality certificate. If the certificate is correct the block (or sequence of blocks) are _postconfirmed_. This requires handling who should send the certificate to the L1 contract, and how to verify the certificate.
+
+
+## Reference Implementation
+
+To simplify we assume that each validator stakes the same amount.
+The set of validators is in charge of validating a block at each round.
+
+A leader validator $V_l$ is elected in each round, and proposes the next transition (and block $B'$):  $B \xrightarrow{\ txs \ } B'$.
+They can do by sending a digest of $txs$ (Merkle root) and a digest of $B'$ (Merke root hash of $B'$), or a _change set_.
+Every other validator checks the validity of $B'$ and send their vote to $V_l$.
+Once $V_l$ has received more than 2/3 of :white_check_mark: for $B'$ they sent the votes (who voted :white_check_mark: and who voted :x:) to the L1 staking contract **and** to the DA.
+
+From there on two separate threads of event occur:
+
+- the DA layer returns an _availability certificate_ that a proof of 2/3 super-majority is available for block $B'$. This step should take O(1) second if we use a fast reliable mempool.
+- the Staking contract on L1 _will_ eventually verify the proof of of 2/3 super-majority (on Ethereum mainnet this should take in the order of 13mins).
```

Comment:
we should be more clear that the frequency of attestations to L1 is much lower than the frequency of attestations on L2.  (for example L2-finality certificates can be distributed every L2-block), while postconfirmations are issued only every X L1-blocks. I would even go further and predefine the range of L2-blocks / number of transactions / L1-timeouts that determine the frequency of postconfirmations

Reply:
yes, we may add that in the optimisations.
Instead of post confirming every new block, confirm sequences of blocks? 
That can be done using a Merkle root for the sequence of blocks, so the design principles are the same.

Reply:
added this to optimisations.